### DISCUSSION TITLE
Race condition on CI image build on post-commit check?

### DISCUSSION THREAD

**Discussion by Sonins** [apache/airflow#17543] [General] [answered]
I think I found some kinda race condition in post-commit check, but I'm not sure. So I want some opinions.

In my opinion, `Build images` task should run before `CI Build` task, 
so CI image can be builded in `Build images` phase and ready for `CI Build` phase.

But in my case, `CI Build` runs first, and `Build images` awaits.
![image](https://user-images.githubusercontent.com/26663842/128957833-44cb7b10-18f3-4749-ab47-4aa1d76972f6.png)

In the result, `Build images` waits for self-hosted runner.

![image](https://user-images.githubusercontent.com/26663842/128959282-482ea8bc-fa93-4bed-add2-13be0f306cd9.png)

And `CI Build` waits for image that cannot be prepared.

![image](https://user-images.githubusercontent.com/26663842/128957941-72fe697f-b47a-4663-93f9-ea314f69dda1.png)

Workflows eventually failed because of timeout.

Is it normal behavior or race condition?

link on actions:

`Build images`: https://github.com/apache/airflow/actions/runs/1118168267
`CI Build` : https://github.com/apache/airflow/actions/runs/1118168312

**Comment by potiuk** [comment, marked as answer]
That was a delay on queue processing (GitHub experienced some issues yesterday)

Unfortunately it cannot be done differently until GitHub introduces "pausing" of workflows waiting for other workflows (which is on a roadmap, but not yet implemented). We need to build and upload images in the "Build image" flow, but this flow can only be triggered as "pull_request_target" (becasue of permissions and sucurity) which can only be triggered AFTER the "pull_request" ("CI build") workflow has been started. Also the "Build images" is "safe" (i.e. you cannot really sneak in some code from fork PR) so we run it on our self-hosted infrastructure rather than on Public GitHub runners (to make it faster in general) 

What we've implemented is that the "CI Build" has a special "wait for images" step which waits (actively) until the "Build Image" one prepares the image. Once Github implements "pausing" workflow and some way of programmatic "unpausing" it, we will be able to use it - but for now this "CI Build" has to actively wait for images. And we cannot wait indefinitely there - so if the images are not there in ~2 hours (I believe), the whole build gets cancelled. I think this is what you experienced.

In your case, apparently we had some delays in building the images recently and as a result you got this delay.

Here you can see full description and detailed sequence diagrams describing our CI architecture - in case you have more questions and doubts how it works.

https://github.com/apache/airflow/blob/main/CI.rst#ci-sequence-diagrams

Also if you have any suggestions how to improve it - PRs are most welcome.

**Reply by Sonins** [reply]
Thanks for detail reply. I doubted that this may be some kind of github problem also, but I was not sure.

Anyway, my PR reopened and is doing check again now.
Thanks for your reply again.

**Reply by potiuk** [reply]
Yeah. Build has different queue for non-committers for security - the "PR" builds use public runner (for committers they share the same queue). And usually the "self-hosted" ones are much faster- but we had a "version" problem which i just fixed and the builds started to pick-up again - hence the "race" you observed.
